### PR TITLE
kvm: add MV Settings for virtual GPU hardware type and memory

### PR DESCRIPTION
**Pull request by leolleeooleo**
### Description

Add MV Settings for virtual GPU hardware type and memory.
The default virtual GPU hardware model type is cirrus.
KVM libvirt are support cirrus, vga, qxl and virtio.
Set the value in MV setting to override /etc/cloudstack/agent/agent.properties
video.hardware=qxl (override vm.video.hardware)
video.ram=32768 (override vm.video.ram)

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

image
image

image
image

### How Has This Been Tested?
I had test windows 10 and ubuntu 20.04 on CloudStack 4.15.2
In my test set "video.hardware=qxl, video.ram=32768" on windows 10 can get better performance.
And set "video.hardware=virtio, video.ram=32768" on ubuntu 20.04 can get better performance.
Windows driver "Stable virtio-win ISO" is from [here](https://github.com/virtio-win/virtio-win-pkg-scripts/blob/master/README.md).

### CODE REVIEW

**Comment by weizhouapache** [comment]
@leolleeooleo 
can you add add the available options ?

like https://github.com/apache/cloudstack/blob/main/server/src/main/java/com/cloud/api/query/QueryManagerImpl.java#L3881
```
        if (HypervisorType.KVM.equals(hypervisorType)) {
            options.put(VmDetailConstants.ROOT_DISK_CONTROLLER, Arrays.asList("osdefault", "ide", "scsi", "virtio"));
        }
```

**Comment by weizhouapache** [comment]
@nvazquez 
this is a good feature. added to 4.16.0.0

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 1]
```diff
@@ -3879,6 +3879,8 @@ private void fillVMOrTemplateDetailOptions(final Map<String, List<String>> optio
 
         if (HypervisorType.KVM.equals(hypervisorType)) {
             options.put(VmDetailConstants.ROOT_DISK_CONTROLLER, Arrays.asList("osdefault", "ide", "scsi", "virtio"));
+            options.put(VmDetailConstants.VIDEO_HARDWARE, Arrays.asList("cirrus", "vga", "qxl", "virtio"));
+            options.put(VmDetailConstants.VIDEO_RAM, Arrays.asList("16384", "32768", "65536"));
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3883]: @leolleeooleo 
can videoram be any integer value ?
if so, you can use
```
options.put(VmDetailConstants.VIDEO_RAM, Collections.emptyList());
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 2]
```diff
@@ -2488,7 +2488,17 @@ protected ConsoleDef createConsoleDef() {
         return new ConsoleDef(PTY, null, null, (short)0);
     }
 
-    protected VideoDef createVideoDef() {
+    protected VideoDef createVideoDef(VirtualMachineTO vmTO) {
+        Map<String, String> details = vmTO.getDetails();
+        if (details != null) {
+            if (details.containsKey(VideoDef.VIDEO_MODEL)) {
+                _videoHw = details.get(VideoDef.VIDEO_MODEL);
```
Comment (weizhouapache) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:2495]: @leolleeooleo 
_videoHw and _videoRam are global variables. I think local variables should be used here.

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/api/query/QueryManagerImpl.java]: @leolleeooleo 
I was not asking you to change the values. sorry for confusion. 
I just wanted to know if there is any restriction here . for example, the min value, the max value, or it must be power of 2.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java (hunk 2 above)
Comment (leolleeooleo) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:2495]: I copy from https://github.com/apache/cloudstack/blob/789d68e43b9e70eecdebbcbf09fcacf25212cc63/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L1040

The value _videoHw and _videoRam are load from /etc/cloudstack/agent/agent.properties
If details value exist will override the value.

Do you mean:
```
    protected VideoDef createVideoDef(VirtualMachineTO vmTO) {
        Map<String, String> details = vmTO.getDetails();
        if (details != null) {
            if (details.containsKey(VideoDef.VIDEO_MODEL)) {
                String videoHw = details.get(VideoDef.VIDEO_MODEL);
                int videoRam = 16384;
                if (details.containsKey(VideoDef.VIDEO_RAM)) {
                    String value = details.get(VideoDef.VIDEO_RAM);
                    videoRam = NumbersUtil.parseInt(value, videoRam);
                }
            return new VideoDef(videoHw, videoRam);
            }
        }
        return new VideoDef(_videoHw, _videoRam);
    }
```

Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:2495]: @leolleeooleo exactly. we should not change the global variables.

you can use 
```
String videoHw = _videoHw;
int videoRam = _videoRam;
if (details != null) {
      ...... // update videoHw and videoRam
}
return new VideoDef(videoHw, videoRam);
```

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 3]
```diff
@@ -2490,16 +2490,18 @@ protected ConsoleDef createConsoleDef() {
...
+        int videoRam = _videoRam;
         if (details != null) {
             if (details.containsKey(VideoDef.VIDEO_MODEL)) {
-                _videoHw = details.get(VideoDef.VIDEO_MODEL);
+                videoHw = details.get(VideoDef.VIDEO_MODEL);
             }
             if (details.containsKey(VideoDef.VIDEO_RAM)) {
                 String value = details.get(VideoDef.VIDEO_RAM);
-                _videoRam = NumbersUtil.parseInt(value, 0);
+                videoRam = NumbersUtil.parseInt(value, 0);
```
Comment (weizhouapache) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:2501]: @leolleeooleo 
better to use
```
videoRam = NumbersUtil.parseInt(value, videoRam);
```

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
manually tested ok.

default
```
    <video>
      <model type='cirrus' vram='16384' heads='1' primary='yes'/>
      <alias name='video0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
    </video>
```

video.hardware=qxl
```
    <video>
      <model type='qxl' ram='65536' vram='65536' vgamem='16384' heads='1' primary='yes'/>
      <alias name='video0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
    </video>
```

video.hardware=qxl, video.ram=32000
```
    <video>
      <model type='qxl' ram='65536' vram='32768' vgamem='16384' heads='1' primary='yes'/>
      <alias name='video0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
    </video>

```

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java (hunk 1 above)
Comment (leolleeooleo) [reply server/src/main/java/com/cloud/api/query/QueryManagerImpl.java]: Note that if VIDEO_RAM = 0, VIDEO_HARDWARE will not work.

If not power of 2, libvirt will auto make it to power of 2.

https://github.com/apache/cloudstack/blob/07f30f6867da2c98e44a61eaf2cfd21c2647725b/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java#L1631

Comment (weizhouapache) [reply server/src/main/java/com/cloud/api/query/QueryManagerImpl.java]: would it better to change to the following ?
```
            if (_videoModel != null && !_videoModel.isEmpty()) {
                videoBuilder.append("<video>\n");
                if (_videoRam != 0) {
                    videoBuilder.append("<model type='" + _videoModel + "' vram='" + _videoRam + "'/>\n");
                } else {
                    videoBuilder.append("<model type='" + _videoModel + "'/>\n");
                }
                videoBuilder.append("</video>\n");
                return videoBuilder.toString();
            }
```

Comment (leolleeooleo) [reply server/src/main/java/com/cloud/api/query/QueryManagerImpl.java]: It is a good idea.
I was worry about it if videoRam is not provided.
I had test it with no error, and here are the document.
https://libvirt.org/formatdomain.html#video-devices
If no value is provided the default is used.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 4]
```diff
@@ -2488,8 +2488,20 @@ protected ConsoleDef createConsoleDef() {
...
     }
 
-    protected VideoDef createVideoDef() {
-        return new VideoDef(_videoHw, _videoRam);
+    protected VideoDef createVideoDef(VirtualMachineTO vmTO) {
+        Map<String, String> details = vmTO.getDetails();
+        String videoHw = _videoHw;
+        int videoRam = _videoRam;
+        if (details != null) {
+            if (details.containsKey(VideoDef.VIDEO_MODEL)) {
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:2496]:
```suggestion
            if (details.containsKey(VmDetailConstants.VIDEO_HARDWARE)) {
```

can use the same constants defined in VmDetailConstants.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 5]
```diff
@@ -2488,8 +2488,20 @@ protected ConsoleDef createConsoleDef() {
...
+                videoHw = details.get(VideoDef.VIDEO_MODEL);
+            }
+            if (details.containsKey(VideoDef.VIDEO_RAM)) {
+                String value = details.get(VideoDef.VIDEO_RAM);
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:2500]:
```suggestion
                String value = details.get(VmDetailConstants.VIDEO_RAM);
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java [hunk 6]
```diff
@@ -1613,6 +1613,10 @@ public String toString() {
     }
 
     public static class VideoDef {
+
+        public static final String VIDEO_MODEL = "video.hardware";
+        public static final String VIDEO_RAM = "video.ram";
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java:1618]: @leolleeooleo you can re-use the constants at VmDetailConstants (to use in LibvirtComputingResource), no need of this constants.

**Review by sureshanaparti — APPROVED** [review]
@leolleeooleo left few comments for minor code improvements, rest of the code LGTM.

**Comment by leolleeooleo** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Pull request merged**